Strange diagnostic message with a suggestion to annotate the lifetime

I wanted to issue a bug report in `bumpalo` and show an example of a vector error message from std
But I got a very strange suggest from compiler:
```rust
fn foo() -> impl Iterator<Item = usize> {
    let arena = Bump::new();
    // vec![in &arena; 1, 2, 3].into_iter()
    let mut vec = Vec::new_in(&arena);
    vec.push(1);
    vec.push(2);
    vec.push(3);
    vec.into_iter()
}
```
[link to playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7aa3f3354b1434f1ff1959690a260eaf)
```
error[E0597]: `arena` does not live long enough
  --> src\main.rs:8:31
   |
8  |     let mut vec = Vec::new_in(&arena);
   |                               ^^^^^^ borrowed value does not live long enough
...
12 |     vec.into_iter()
   |     --------------- opaque type requires that `arena` is borrowed for `'static`
13 | }
   | - `arena` dropped here while still borrowed
   |
help: you can add a bound to the opaque type to make it last less than `'static` and match `'static`
   |
12 |     vec.into_iter() + 'static
   |                     +++++++++
```
I even thought there was progress in [`type_ascription`](https://rust-lang.github.io/rfcs/0803-type-ascription.html) feature
In general, this message is very strange. Right now in growth I can't make arena live longer (I can, but not through annotations)

--- a/compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs
@@ -273,13 +273,17 @@ impl<'tcx> BorrowExplanation<'tcx> {
             _ => {}
         }
     }
-    pub(crate) fn add_lifetime_bound_suggestion_to_diagnostic(
+
+    fn add_lifetime_bound_suggestion_to_diagnostic(
         &self,
         err: &mut Diagnostic,
         category: &ConstraintCategory<'tcx>,
         span: Span,
         region_name: &RegionName,
     ) {
+        if !span.is_desugaring(DesugaringKind::OpaqueTy) {
+            return;
+        }
         if let ConstraintCategory::OpaqueType = category {
             let suggestable_name =
                 if region_name.was_named() { region_name.name } else { kw::UnderscoreLifetime };

--- a/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.rs
+++ b/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.rs
@@ -1,0 +1,29 @@
+#![feature(allocator_api)]
+
+use std::{
+    alloc::{AllocError, Allocator, Layout},
+    ptr::NonNull,
+};
+
+struct GhostBump;
+
+unsafe impl Allocator for &GhostBump {
+    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        todo!()
+    }
+
+    unsafe fn deallocate(&self, ptr: NonNull<u8>, layout: Layout) {
+        todo!()
+    }
+}
+
+fn foo() -> impl Iterator<Item = usize> {
+    let arena = GhostBump;
+    let mut vec = Vec::new_in(&arena); //~ ERROR `arena` does not live long enough
+    vec.push(1);
+    vec.push(2);
+    vec.push(3);
+    vec.into_iter()
+}
+
+fn main() {}

--- a/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.rs
+++ b/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.rs
@@ -1,29 +1,12 @@
-#![feature(allocator_api)]
+pub trait T {}
 
-use std::{
-    alloc::{AllocError, Allocator, Layout},
-    ptr::NonNull,
-};
+struct S<'a>(&'a ());
 
-struct GhostBump;
+impl<'a> T for S<'a> {}
 
-unsafe impl Allocator for &GhostBump {
-    fn allocate(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
-        todo!()
-    }
-
-    unsafe fn deallocate(&self, ptr: NonNull<u8>, layout: Layout) {
-        todo!()
-    }
-}
-
-fn foo() -> impl Iterator<Item = usize> {
-    let arena = GhostBump;
-    let mut vec = Vec::new_in(&arena); //~ ERROR `arena` does not live long enough
-    vec.push(1);
-    vec.push(2);
-    vec.push(3);
-    vec.into_iter()
+fn foo() -> impl T {
+    let x = ();
+    S(&x) //~ ERROR `x` does not live long enough
 }
 
 fn main() {}

--- a/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.stderr
+++ b/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.stderr
@@ -1,0 +1,14 @@
+error[E0597]: `arena` does not live long enough
+  --> $DIR/do-not-suggest-adding-bound-to-opaque-type.rs:22:31
+   |
+LL |     let mut vec = Vec::new_in(&arena);
+   |                               ^^^^^^ borrowed value does not live long enough
+...
+LL |     vec.into_iter()
+   |     --------------- opaque type requires that `arena` is borrowed for `'static`
+LL | }
+   | - `arena` dropped here while still borrowed
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0597`.

--- a/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.stderr
+++ b/src/test/ui/regions/do-not-suggest-adding-bound-to-opaque-type.stderr
@@ -1,13 +1,13 @@
-error[E0597]: `arena` does not live long enough
-  --> $DIR/do-not-suggest-adding-bound-to-opaque-type.rs:22:31
+error[E0597]: `x` does not live long enough
+  --> $DIR/do-not-suggest-adding-bound-to-opaque-type.rs:9:7
    |
-LL |     let mut vec = Vec::new_in(&arena);
-   |                               ^^^^^^ borrowed value does not live long enough
-...
-LL |     vec.into_iter()
-   |     --------------- opaque type requires that `arena` is borrowed for `'static`
+LL |     S(&x)
+   |     --^^-
+   |     | |
+   |     | borrowed value does not live long enough
+   |     opaque type requires that `x` is borrowed for `'static`
 LL | }
-   | - `arena` dropped here while still borrowed
+   | - `x` dropped here while still borrowed
 
 error: aborting due to previous error
 

